src/models/audio.py

```diff
@@ -0,0 +1,79 @@
+import torch
+import numpy as np
+from typing import List, Union
+from .utils import get_mask, masked_emb_pooling
+from transformers import WhisperModel, WhisperFeatureExtractor, AutoProcessor, SeamlessM4TModel
+
+
+class BaseAudioEncoder(torch.nn.Module):
+    def __init__(
+            self,
+            proj_dim: int,
+            device: str = "cpu"
+    ):
+        super().__init__()
+        self.proj_dim = proj_dim
+        self.device = device
+
+    def forward(self, audio: Union[List[np.array], np.array]) -> torch.Tensor:
+        pass
+
+
+class WhisperEncoder(BaseAudioEncoder):
+    def __init__(
+            self,
+            proj_dim: int,
+            model_dim: int = 384,
+            checkpoint: str = "openai/whisper-tiny",
+            device: str = "cpu"
+    ):
+        super().__init__(proj_dim, device)
+        self.model = WhisperModel.from_pretrained(checkpoint).encoder.to(self.device)
+        self.extractor = WhisperFeatureExtractor.from_pretrained(checkpoint)
+        self.projection_layer = torch.nn.Linear(model_dim, self.proj_dim).ti(self.device)
+
+    def forward(self, audio: Union[List[np.array], np.array]) -> torch.Tensor:
+        feats = self.extractor(
+            audio,
+            padding=True,
+            sampling_rate=16000,
+            return_tensors="pt",
+            return_attention_mask=True
+        )
+        feats = {k: v.to(self.device) for k, v in feats.items()}
+        encoder_output = self.model(**feats).last_hidden_state
+        mask = get_mask(feats["attention_mask"], encoder_output.shape[1])
+        embeddings = masked_emb_pooling(mask, encoder_output)
+        if isinstance(audio, np.ndarray):
+            embeddings = embeddings[0]
+        embeddings = self.projection_layer(embeddings)
+        return embeddings
+
+
+class SeamlessFacebookEncoder(BaseAudioEncoder):
+    def __init__(
+            self,
+            proj_dim: int,
+            model_dim: int = 1024,
+            checkpoint: str = "facebook/hf-seamless-m4t-medium",
+            device: str = "cpu"
+    ):
+        super().__init__(proj_dim, device)
+        self.model = SeamlessM4TModel.from_pretrained(checkpoint).speech_encoder.to(self.device)
+        self.extractor = AutoProcessor.from_pretrained(checkpoint)
+        self.projection_layer = torch.nn.Linear(model_dim, self.proj_dim).to(self.device)
+
+    def forward(self, audio: Union[List[np.array], np.array]) -> torch.Tensor:
+        feats = self.extractor(
+            audios=audio,
+            sampling_rate=16000,
+            return_tensors="pt",
+        )
+        feats = {k: v.to(self.device) for k, v in feats.items()}
+        encoder_output = self.model(**feats).last_hidden_state
+        mask = get_mask(feats["attention_mask"], encoder_output.shape[1])
+        embeddings = masked_emb_pooling(mask, encoder_output)
+        if isinstance(audio, np.ndarray):
```

Comment:
what does it do

Reply:
there are documentation in get_mask and masked_emb_pooling  functions, but in a few words it transforms attention mask from feature extractor (so that the dims are correct) and use this mask to mask and pool embeddings

Reply:
No I mean why do we have if is instance(audio, np.ndarray)

Reply:
I have added it in case you want to find embedding of single audio